sdk/core/switcher/entry.S

```diff
@@ -91,11 +91,21 @@ compartment_switcher_entry:
 	// XXX: This line is useless, only for mscratch to show up in debugging.
 	cmove			ct2, ct2
 #endif
-	clhu			tp, TrustedStack_offset_frameoffset(ct2)
+
 	// make sure the trusted stack is still in bounds
-	cgetlen                 t2, ct2
-	bgeu                    tp, t2, .Lout_of_trusted_stack
+	clhu			tp, TrustedStack_offset_frameoffset(ct2)
+	cgetlen         t2, ct2
+	bgeu           	tp, t2, .Lout_of_trusted_stack
+	// make sure the caller's CSP is valid
+	cgettag         t2, csp
+	beqz            t2, .Lforce_unwind
+	// make sure the caller's CSP has the expected permissions
+	cgetperm        t2, csp
+	li              tp, 0x7e
+	bne	            tp, t2, .Lforce_unwind
```

Comment:
Can these be moved before we access mscratch?  We don't need access to any privileged state to do these checks.

Reply:
True, great point. I moved it to the beginning of the trap